src/image_service.rs

```diff
@@ -1,24 +1,195 @@
 use crate::criapi::{self, image_service_server::ImageService};
-use std::collections::HashMap;
+use crate::{criapi::ImageSpec, lock_map::LockMap};
+use anyhow::Result;
+use async_compression::tokio_02::bufread::GzipDecoder;
+use bytes::Bytes;
+use oci_registry_client::{
+    blob::Blob, manifest::Digest, manifest::Manifest, DockerRegistryClientV2,
+};
+use prost::Message;
+use std::path::{Path, PathBuf};
+use std::{collections::HashMap, sync::Arc};
+use tokio::{fs, io::stream_reader, stream::Stream, sync::Mutex};
+use tokio_tar::Archive;
 use tonic::{Request, Response, Status};
 
-#[derive(Default)]
-pub struct MyImage {}
+#[derive(thiserror::Error, Debug)]
+pub enum ImageError {
+    #[error("image database error: {0}")]
+    DatabaseError(#[from] sled::Error),
+    #[error("decode error, database likely corrupted: {0}")]
+    DecodeError(#[from] prost::DecodeError),
+}
+
+impl From<ImageError> for Status {
+    fn from(_: ImageError) -> Self {
+        todo!()
+    }
+}
+
+pub struct MyImage {
+    database: sled::Db,
+    layers: PathBuf,
+
+    layer_lock: LockMap<String>,
+
+    registries: Vec<DockerRegistryClientV2>,
+    // TODO: prevent pulling same image simultaneously
+    // pull_progress: RwLock<HashMap<ImageSpec, Mutex<()>>>,
+}
+
+/// Workaround for bad original API
+/// Blob should implement AsyncRead itself
+fn blob_to_stream(
+    mut blob: Blob,
+    out_digest: Arc<Mutex<Option<Digest>>>,
+) -> impl Stream<Item = tokio::io::Result<Bytes>> {
+    async_stream::try_stream! {
+        while let Some(bytes) = blob.chunk().await.map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))? {
+            yield bytes;
+        }
+        out_digest.lock().await.replace(blob.digest());
+    }
+}
+
+impl MyImage {
+    /// Open directory as image storage
+    pub async fn open(image_storage: &Path) -> Result<Self> {
+        let mut db_path = image_storage.to_owned();
+        db_path.push("images.db");
+        let mut layers = image_storage.to_owned();
+        layers.push("layers");
+        fs::create_dir_all(&layers).await?;
+        Ok(MyImage {
+            database: sled::open(db_path)?,
+            layers,
+            registries: vec![DockerRegistryClientV2::new(
```

Comment:
Maybe instead of hard coding the registries here maybe in future we can read from the config file, just like how we do it [currently](https://github.com/cri-o/cri-o/blob/master/pkg/config/config.go#L349) in CRI-O.